meetings/working-groups/collection-literals/CL-2023-08-03.md

```diff
@@ -0,0 +1,140 @@
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c];        // heap allocated, global scope.
+scoped ReadonlySpan<int> x2 = [a, b, c]; // stack allocated, local scope.
+```
+
+This has the positive that both heap and local allocations are possible, and choosing between them is as simple as adding `scoped` or not.  Unfortunately it has several major negatives the working group was quite dissatisfied with:
+
+1. We believe the *common* case for using spans with collection expressions will be for people wanting stack allocated data.  Having the default (non-`scoped`) syntax basically flips that and makes the common case more cluttered and less pleasant to work with.
+2. We believe the standard intuition for nearly all users (including experts in the low-level) space would be that the first line produces stack-allocated data.  After all, that's the most common reason to create locals of span types, and their belief would be that "of course if i initialized with a collection expression" that "it will do the right thing".  
+3. We believe that the first line would actually be a *large* "pit of fail" for users.  It would be very easy and attractive for customers to just take existing code, hear about collection expressions (including that it works for spans) and translate `stackalloc` code over to just using the collection expression.  This would then result in near immediate performance tanking due to potentially extreme allocations happening in hotspots that didn't exist before.  Such problems could even easily lead to this feature being entirely banned in codebases due to the fear of this happening.
+4. We believe if we had this behavior we would have to have a diagnostic (likely on by default) that then immediately warned the user that they were getting a heap allocation on 'x1' and to either:
+    1. suppress the diagnostic if they really wanted that, or
+    2. put an explicit cast (e.g. `(int[])[a, b, c]`) to indicate that they indeed did want this behavior, or
+    3. add the `scoped` modifier.
+
+All in all, while this allowed for both cases to easily be distinguished with a simple modifier, everything about the defaults and behavior was flipped from where we wanted it to be.  This then led us to option 2:
+
+### Option 2. Determine what scoping a target has based on usage.
+
+A discussion has been going on about allowing the user to just write:
+
+```c#
+ReadOnlySpan<int> x1 = [a, b, c]; // Could be local or global scoped
+scoped ReadOnlySpan<int> x2 = [a, b, c]; // Always local scoped.
+```
+
+and then determining the scoping based on how the variable is used.  For example, if 'x1' does not escape, then it could be stack allocated.  If it does end up escaping though (for example with `return x1;`), the collection expression would be heap allocated.  'x2' would always be local scoped of course and would be prevented from escaping.
```

Comment:
I believe the discussion was to use the static scope of the target, independent of how the target is used.
```c#
ReadOnlySpan<int> x1 = stackalloc int[1]; // local scope
...
x1 = [1, 2, 3];  // collection expression has local scope
```